File: bots/rs/drewfett/v1000/src/builder/tasks/econ/infrastructure/resolve_congestion.rs
//! Translation of `bots/intgrah/v54.7.9/builder/tasks/econ/infrastructure/resolve_congestion.py`.
//!
//! Relieve an overcapacity junction by destroying one of its immediate
//! friendly feeders. The removed feeder's upstream chain becomes dangling
//! at the old feeder tile; subsequent extend-chain tasks reroute it to a
//! less-saturated sink. Candidate junctions come from `congested_junctions`
//! (populated empirically by `update_economy_reachability`).

use cambc::{Controller, ControllerApi, Position};
use serde_json::Map;

use crate::builder::Builder;
use crate::builder::helpers::make_move;
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};
use crate::util::debug::debug as log;
use crate::util::metrics::chebyshev;
use crate::util::posint::{PosInt, idx_of};

pub fn resolve_congestion(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    if pyrust::vec::is_empty!(self_.congested_junctions) {
        return Some(TaskRejected::new("no congested junction in range"));
    }
    log(
        &format!(
            "resolve_congestion: {} congested junctions visible",
            pyrust::len!(self_.congested_junctions)
        ),
        Map::new(),
    );

    // Sort the junction iteration so candidate-feeder accumulation is
    // deterministic across hash-randomized iteration of `congested_junctions`.
    // PosInt = y*100+x, so sorting by PosInt is identical to sorting by (y,x).
    let mut junctions: Vec<PosInt> =
        pyrust::collect!(pyrust::copied!(pyrust::iter!(self_.congested_junctions)));
    pyrust::sort!(junctions);
    let mut targets: Vec<Position> = pyrust::vec::new!();
    for j_pi in junctions {
        for &feeder in &self_.in_edges[j_pi as usize] {
            let fi = idx_of(feeder) as usize;
            if pyrust::is_none!(self_.building_kind[fi]) {
                continue;
            }
            if self_.building_team[fi] != Some(self_.my_team) {
                continue;
            }
            pyrust::vec::push!(targets, feeder);
        }
    }
    // Same feeder may be reachable through multiple junctions; dedupe and
    // re-sort so destroy / approach order is fully deterministic.
    pyrust::sort_by_key!(targets, |p| (p.y, p.x));
    // Remove consecutive duplicates after sorting.
    let mut targets_dedup: Vec<Position> = pyrust::vec::new!();
    let mut _prev_dedup: Option<Position> = None;
    for _p_dedup in pyrust::iter!(targets) {
        if _prev_dedup != Some(*_p_dedup) {
            pyrust::vec::push!(targets_dedup, *_p_dedup);
        }
        _prev_dedup = Some(*_p_dedup);
    }
    let targets = targets_dedup;

    if pyrust::vec::is_empty!(targets) {
        log(
            "resolve_congestion: no friendly feeders to remove",
            Map::new(),
        );
        return Some(TaskRejected::new(
            "congested junction has no friendly feeder to remove",
        ));
    }
    log(
        &format!(
            "resolve_congestion: {} candidate feeders",
            pyrust::len!(targets)
        ),
        Map::new(),
    );

    for &feeder in &targets {
        if pyrust::unwrap!(ct.can_destroy(feeder)) {
            log(
                &format!("resolve_congestion: DESTROY feeder {feeder:?}"),
                Map::new(),
            );
            pyrust::unwrap!(ct.destroy(feeder));
            self_.apply_local_destroy(feeder);
            return None;
        }
    }

    let my_pos = self_.my_pos;
    // Tiebreak by (y, x) on top of chebyshev distance.
    let nearest = *pyrust::unwrap!(pyrust::min_by!(pyrust::iter!(targets), |&&p| (
        chebyshev(my_pos, p),
        p.y,
        p.x
    )));
    log(
        &format!("resolve_congestion: walking toward nearest feeder {nearest:?}"),
        Map::new(),
    );
    if make_move(self_, ct, nearest) {
        return None;
    }
    log(
        &format!("resolve_congestion: could not approach {nearest:?}"),
        Map::new(),
    );
    Some(TaskRejected::from_string(format!(
        "cannot approach feeder {nearest:?}"
    )))
}

File: bots/beacon/bestrn/builder/tasks/econ/infrastructure/resolve_congestion.py
"""
Translation of `bots/intgrah/v54.7.9/builder/tasks/econ/infrastructure/resolve_congestion.py`.

Relieve an overcapacity junction by destroying one of its immediate
friendly feeders. The removed feeder's upstream chain becomes dangling
at the old feeder tile; subsequent extend-chain tasks reroute it to a
less-saturated sink. Candidate junctions come from `congested_junctions`
(populated empirically by `update_economy_reachability`).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cambc import Position
from util.debug import debug as log
from util.metrics import chebyshev

from builder.helpers import make_move
from builder.tasks.rejected import TaskRejected


def resolve_congestion(self_, ct):
    if not self_.congested_junctions:
        return TaskRejected("no congested junction in range")
    log(
        f"resolve_congestion: {len(self_.congested_junctions)} congested junctions visible",
        {},
    )
    junctions: list[Position] = list(self_.congested_junctions)
    junctions.sort(key=lambda p: (p.y, p.x))
    targets: list[Position] = []
    for j in junctions:
        for feeder in self_.in_edges[int(j.y) * 50 + int(j.x)]:
            fi = int(feeder.y) * 50 + int(feeder.x)
            if self_.building_kind[fi] is None:
                continue
            if self_.building_team[fi] != self_.my_team:
                continue
            targets.append(feeder)
    targets.sort(key=lambda p: (p.y, p.x))
    targets.__setitem__(
        slice(None),
        [__x for __i, __x in enumerate(targets) if __i == 0 or __x != targets[__i - 1]],
    )
    if not targets:
        log("resolve_congestion: no friendly feeders to remove", {})
        return TaskRejected("congested junction has no friendly feeder to remove")
    log(f"resolve_congestion: {len(targets)} candidate feeders", {})
    for feeder in targets:
        if ct.can_destroy(feeder):
            log(f"resolve_congestion: DESTROY feeder {feeder!r}", {})
            ct.destroy(feeder)
            self_.apply_local_destroy(feeder)
            return None
    my_pos = self_.my_pos
    nearest = (
        min(targets, key=lambda p: (chebyshev(my_pos, p), p.y, p.x))
        if targets
        else None
    )
    log(f"resolve_congestion: walking toward nearest feeder {nearest!r}", {})
    if make_move(self_, ct, nearest):
        return None
    log(f"resolve_congestion: could not approach {nearest!r}", {})
    return TaskRejected.from_string(f"cannot approach feeder {nearest!r}")
